vowpalwabbit/core/include/vw/core/reductions/eigen_memory_tree.h

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#pragma once
+#include "vw/core/example.h"
+#include "vw/core/rand_state.h"
+
+#include <list>
+#include <unordered_map>
+#include <vector>
+
+namespace VW
+{
+namespace reductions
+{
+VW::LEARNER::base_learner* eigen_memory_tree_setup(VW::setup_base_i& stack_builder);
+
+namespace eigen_memory_tree
+{
+using emt_feats = std::vector<std::pair<int, float>>;
+
+enum class emt_scorer_type
+{
+  random = 1,
+  distance = 2,
+  self_consistent_rank = 3,
+  not_self_consistent_rank = 4
+};
+
+enum class emt_router_type
+{
+  random = 1,
+  eigen = 2
+};
+
+float emt_median(std::vector<float>&);
+
+float emt_inner(const emt_feats&, const emt_feats&);
+
+float emt_norm(const emt_feats&);
+
+void emt_abs(emt_feats&);
+
+void emt_scale(emt_feats&, float);
+
+void emt_normalize(emt_feats&);
+
+emt_feats emt_scale_add(float, const emt_feats&, float, const emt_feats&);
+
+emt_feats emt_router_eigen(std::vector<emt_feats>&, VW::rand_state&);
+
+template <typename RandomIt>
+void emt_shuffle(RandomIt first, RandomIt last, VW::rand_state* rng)
```

Comment:
We expect this to be non-null and we don't check for null, so we should use a reference instead.